### PR TITLE
Native methods

### LINKED ISSUE #1028: How to perform hit test?
# Question

From javascript, how can we perform hit test to find out if the specified (x, y) coordinates fall within an SVG element?

Obviously this functionality exists somewhere or the various onPressXXX() events would not work. I am just not able to figure out how.

### PR DESCRIPTION
**Pull request by msand**
# Summary
Initial implementations of:
https://svgwg.org/svg2-draft/coords.html#BoundingBoxes
https://svgwg.org/svg2-draft/types.html#SVGBoundingBoxOptions
https://svgwg.org/svg2-draft/types.html#InterfaceSVGGraphicsElement
https://svgwg.org/svg2-draft/types.html#__svg__SVGGraphicsElement__getBBox
https://svgwg.org/svg2-draft/types.html#__svg__SVGGraphicsElement__getCTM
https://svgwg.org/svg2-draft/types.html#__svg__SVGGraphicsElement__getScreenCTM

https://svgwg.org/svg2-draft/types.html#InterfaceSVGGeometryElement
https://svgwg.org/svg2-draft/types.html#__svg__SVGGeometryElement__getTotalLength
https://svgwg.org/svg2-draft/types.html#__svg__SVGGeometryElement__getPointAtLength
https://svgwg.org/svg2-draft/types.html#__svg__SVGGeometryElement__isPointInFill
https://svgwg.org/svg2-draft/types.html#__svg__SVGGeometryElement__isPointInStroke

https://svgwg.org/specs/paths/#InterfaceSVGPathElement
https://svgwg.org/specs/paths/#__svg__SVGPathElement__getTotalLength
https://svgwg.org/specs/paths/#__svg__SVGPathElement__getPointAtLength

https://www.w3.org/TR/SVG11/types.html#InterfaceSVGLocatable
https://www.w3.org/TR/SVG11/types.html#__svg__SVGLocatable__getBBox
https://www.w3.org/TR/SVG11/types.html#__svg__SVGLocatable__getCTM
https://www.w3.org/TR/SVG11/types.html#__svg__SVGLocatable__getScreenCTM

https://www.w3.org/TR/SVG11/paths.html#InterfaceSVGPathElement
https://www.w3.org/TR/SVG11/paths.html#__svg__SVGPathElement__getTotalLength
https://www.w3.org/TR/SVG11/paths.html#__svg__SVGPathElement__getPointAtLength

isPointInFill - Fixes #1028
isPointInStroke
getTotalLength
getPointAtLength
getBBox
getCTM
getScreenCTM

<https://github.com/react-native-community/react-native-svg/pull/1135/files#diff-0e5b6a37c9af5cccba5b9fef9b58006cR58-R117>

# TODO
Verify compatibility with web / conformance with spec

### CODE REVIEW

**Comment by msand** [comment]
@shergin would you mind having a look at this? Is there a better way to make calls to native methods I should be using here?

**Comment by msand** [comment]
@EvanBacon @kristerkari Would you mind having a look / testing these out?

**Comment by kristerkari** [comment]
Sure, I can have a look and help with testing.

**Comment by msand** [comment]
@msageryd Perhaps you would have use for some of these as well?

**Comment by msageryd** [comment]
@msand Wow, this is really cool. When I have time I'll try to switch out my home-baked getBBox.

I haven't used BBox in *real* svg, but my understanding is that it is calculated before any transforms. If I want the boundingbox of a scaled or rotated object, would I need to take any extra steps via getCTM and manually transform the bounding box with the matrix from the object?

I'm sorry if this is a stupid question. As I said, I'm not a web guy. The only real SVG experience I have had is via react-native-svg.

**Comment by msageryd** [comment]
@msand It seems from your sample code above that marker support is now implemented. Is that released as well? Really cool in that case. I'm not sure how I missed these goodies..

**Comment by msand** [comment]
@msageryd What remains to do with these new methods is to ensure they behave similarly enough to the web ones. 

Correct, getCTM * getBBox should give you the bounds in the coordinate system of the Svg viewport, as you describe.

Marker, context-stroke, context-fill and related onPress/gestures support was finished essentially yesterday. So they're fresh out of the oven as well ;) Any help with testing those / uncovering bugs is greatly appreciated as well.

**Comment by msand** [comment]
And also, if anyone has any requests for more native methods, this would be a good time to mention them.

File: ios/Text/RNSVGTSpan.m [hunk 1]
```diff
@@ -1061,24 +1030,18 @@ + (CGFloat)getTextAnchorOffset:(RNSVGTextAnchor)textAnchor width:(CGFloat) width
...
         if ([parent class] == [RNSVGTextPath class]) {
             textPath = (RNSVGTextPath*) parent;
-            [textPath getPathLength:&_pathLength
-                          lineCount:&lineCount
-                            lengths:&lengths
-                              lines:&lines
-                           isClosed:&isClosed];
+            RNSVGNode *template = [self.svgView getDefinedTemplate:textPath.href];
+            CGPathRef path = [template getPath:nil];
+            [measure extractPathData:path];
```
Comment (msand) [inline ios/Text/RNSVGTSpan.m]: Need to reset measure if transitioning from having a textPath to not having one.

File: ios/Utils/RNSVGPathMeasure.m [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+- (void)addLine:(CGPoint *)last next:(const CGPoint *)next {
+    NSArray *line = @[[NSValue valueWithCGPoint:*last], [NSValue valueWithCGPoint:*next]];
+    _pathLength += distance(*last, *next);
+    [_lengths addObject:[NSNumber numberWithDouble:_pathLength]];
+    [_lines addObject:line];
+    *last = *next;
+}
+
+- (void)extractPathData:(CGPathRef)path {
+    CGPoint origin = CGPointMake (0.0, 0.0);
```
Comment (msand) [inline ios/Utils/RNSVGPathMeasure.m]: return early if path is same as previous

**Comment by msand** [comment]
Web version:


```jsx
const ForwardPath = forwardRef((props, ref) => {
  return <Path {...props} forwardedRef={ref} />;
});

class SvgTest extends React.Component {
  refPath = createRef();
  render() {
    return (
      <TouchableOpacity
        style={styles.container}
        onPress={() => {
          const options = { x: 139, y: 58 };
          const element = this.refPath.current;
          const point = element.ownerSVGElement.createSVGPoint();
          point.x = options.x;
          point.y = options.y;
          const res = this.refPath.current.isPointInFill(point);
          alert(`isPointInFill: ${JSON.stringify(options)} ${res}`);
        }}
      >
        <Svg
          xmlns="http://www.w3.org/2000/svg"
          width="275"
          height="200"
          viewBox="0 0 275 200"
        >
          <Defs>
            <Marker
              id="Triangle"
              viewBox="0 0 10 10"
              refX="1"
              refY="5"
              markerUnits="strokeWidth"
              markerWidth="4"
              markerHeight="3"
              orient="auto"
            >
              <Path d="M 0 0 L 10 5 L 0 10 z" fill="context-stroke" />
            </Marker>
          </Defs>

          <G fill="none" strokeWidth="10">
            <ForwardPath
              ref={this.refPath}
              stroke="crimson"
              markerEnd="url(#Triangle)"
              d="M 100,75 C 125,50 150,50 175,75"
            />
          </G>
        </Svg>
      </TouchableOpacity>
    );
  }
}
```

**Comment by msand** [comment]
This seems to work cross-platform:

App.js
```jsx
import React, {createRef, forwardRef} from 'react';
import {StyleSheet, TouchableOpacity, Platform} from 'react-native';
import {Svg, Defs, Marker, Path, G} from 'react-native-svg';

const isWeb = Platform.OS === 'web';

const ForwardPath = forwardRef((props, ref) => {
  return <Path {...props} forwardedRef={ref} />;
});

const PlatformPath = isWeb ? ForwardPath : Path;

function createPoint(element, options) {
  if (!isWeb) {
    return options;
  }
  const point = element.ownerSVGElement.createSVGPoint();
  point.x = options.x;
  point.y = options.y;
  return point;
}

function alertResults(
  options,
  isPointInFill,
  isPointInStroke,
  totalLength,
  pointAtHalfLength,
  BBox,
  CTM,
  screenCTM,
) {
  const mKeys = ['a', 'b', 'c', 'd', 'e', 'f'];
  const boxKeys = ['x', 'y', 'width', 'height'];
  const message = `isPointInFill: ${JSON.stringify(options)} ${isPointInFill}
isPointInStroke: ${JSON.stringify(options)} ${isPointInStroke}
totalLength: ${totalLength}
pointAtHalfLength: ${pointAtHalfLength.x} ${pointAtHalfLength.y}
BBox: ${boxKeys.map(k => `${k}: ${BBox[k]}`)}
CTM: ${mKeys.map(k => `${k}: ${CTM[k]}`)}
screenCTM: ${mKeys.map(k => `${k}: ${screenCTM[k]}`)}`;
  alert(message);
}

async function testNativeMethods(element) {
  const notInStroke = {x: 168, y: 85};
  const insideStrokeAndFill = {x: 138, y: 58};
  const testStroke = true;
  const option = testStroke ? notInStroke : insideStrokeAndFill;
  const point = createPoint(element, option);
  const CTM = await element.getCTM();
  const BBox = await element.getBBox();
  const screenCTM = await element.getScreenCTM();
  const totalLength = await element.getTotalLength();
  const pointAtHalfLength = await element.getPointAtLength(
    totalLength / 2,
  );
  const isPointInFill = await element.isPointInFill(point);
  const isPointInStroke = await element.isPointInStroke(point);
  alertResults(
    point,
    isPointInFill,
    isPointInStroke,
    totalLength,
    pointAtHalfLength,
    BBox,
    CTM,
    screenCTM,
  );
}

export default class App extends React.Component {
  refPath = createRef();
  render() {
    return (
      <TouchableOpacity
        style={styles.container}
        onPress={() => {
          testNativeMethods(this.refPath.current);
        }}>
        <Svg
          xmlns="http://www.w3.org/2000/svg"
          width="275"
          height="200"
          viewBox="0 0 275 200">
          <Defs>
            <Marker
              id="Triangle"
              viewBox="0 0 10 10"
              refX="1"
              refY="5"
              markerUnits="strokeWidth"
              markerWidth="4"
              markerHeight="3"
              orient="auto">
              <Path d="M 0 0 L 10 5 L 0 10 z" fill="context-stroke" />
            </Marker>
          </Defs>

          <G fill="none" strokeWidth="10" markerEnd="url(#Triangle)">
            <PlatformPath
              ref={this.refPath}
              stroke="crimson"
              d="M 100,75 C 125,50 150,50 175,75"
              markerEnd="url(#Triangle)"
            />
          </G>
        </Svg>
      </TouchableOpacity>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
  },
});

```

**Comment by msand** [comment]
I've improved the getScreenCTM calculation logic in the latest commit, but not completely sure if it's entirely correct yet. Any help in testing / uncovering bugs is much appreciated.

**Comment by msand** [comment]
I've implemented SVGPoint and SVGMatrix now, such that the familiar web api signatures can be used: SVGMatrix.inverse(), SVGPoint.matrixTransform(SVGMatrix) etc. ownerSVGElement available on elements, exposing createSVGPoint and createSVGMatrix, (only have to add await before the async native methods, which are sync on the web but promise returning in react-native):

```jsx
import React, {createRef, forwardRef} from 'react';
import {StyleSheet, TouchableOpacity, Platform} from 'react-native';
import {Svg, Defs, Marker, Path, G} from 'react-native-svg';

const ForwardPath = forwardRef((props, ref) => {
  return <Path {...props} forwardedRef={ref} />;
});

const isWeb = Platform.OS === 'web';
const PlatformPath = isWeb ? ForwardPath : Path;

function createPoint(element, options) {
  const point = element.ownerSVGElement.createSVGPoint();
  point.x = options.x;
  point.y = options.y;
  return point;
}

function alertResults(
  options,
  isPointInFill,
  isPointInStroke,
  totalLength,
  pointAtHalfLength,
  BBox,
  CTM,
  screenCTM,
  inverse,
  screenPoint,
) {
  const mKeys = ['a', 'b', 'c', 'd', 'e', 'f'];
  const boxKeys = ['x', 'y', 'width', 'height'];
  const message = `isPointInFill: ${JSON.stringify(options)} ${isPointInFill}
isPointInStroke: ${JSON.stringify(options)} ${isPointInStroke}
totalLength: ${totalLength}
pointAtHalfLength: ${pointAtHalfLength.x} ${pointAtHalfLength.y}
BBox: ${boxKeys.map(k => `${k}: ${BBox[k]}`)}
CTM: ${mKeys.map(k => `${k}: ${CTM[k]}`)}
screenCTM: ${mKeys.map(k => `${k}: ${screenCTM[k]}`)}
inverse: ${mKeys.map(k => `${k}: ${inverse[k]}`)}
screenPoint: ${screenPoint.x} ${screenPoint.y}`;
  alert(message);
}

async function testNativeMethods(element) {
  const notInStroke = {x: 168, y: 85};
  const insideStrokeAndFill = {x: 138, y: 58};
  const testStroke = false;
  const option = testStroke ? notInStroke : insideStrokeAndFill;
  const point = createPoint(element, option);
  const CTM = await element.getCTM();
  const BBox = await element.getBBox();
  const screenCTM = await element.getScreenCTM();
  const inverse = screenCTM.inverse();
  const screenPoint = point.matrixTransform(inverse);
  const totalLength = await element.getTotalLength();
  const pointAtHalfLength = await element.getPointAtLength(totalLength / 2);
  const isPointInFill = await element.isPointInFill(point);
  const isPointInStroke = await element.isPointInStroke(point);
  alertResults(
    point,
    isPointInFill,
    isPointInStroke,
    totalLength,
    pointAtHalfLength,
    BBox,
    CTM,
    screenCTM,
    inverse,
    screenPoint,
  );
}

export default class App extends React.Component {
  refPath = createRef();
  render() {
    return (
      <TouchableOpacity
        style={styles.container}
        onPress={() => {
          testNativeMethods(this.refPath.current);
        }}>
        <Svg
          xmlns="http://www.w3.org/2000/svg"
          width="275"
          height="200"
          viewBox="0 0 275 200">
          <Defs>
            <Marker
              id="Triangle"
              viewBox="0 0 10 10"
              refX="1"
              refY="5"
              markerUnits="strokeWidth"
              markerWidth="4"
              markerHeight="3"
              orient="auto">
              <Path d="M 0 0 L 10 5 L 0 10 z" fill="context-stroke" />
            </Marker>
          </Defs>

          <G fill="none" strokeWidth="10" markerEnd="url(#Triangle)">
            <PlatformPath
              ref={this.refPath}
              stroke="crimson"
              d="M 100,75 C 125,50 150,50 175,75"
              markerEnd="url(#Triangle)"
            />
          </G>
        </Svg>
      </TouchableOpacity>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
  },
});

```

**Comment by msand** [comment]
Made the native methods synchronous, same as on the web, so no need for await anymore.

**Comment by msand** [comment]
@necolas Would you have any feedback on this? Any specific methods you'd like available in native which you're using on the web? Or any constraints on behavior / what to make available? Especially with regards to react-native-web projects using react-native-svg components, and cross-platform libraries

**Comment by necolas** [comment]
Thanks for the ping. I don't have anything helpful to add as we mainly used SVG for creating icons, which doesn't need much of the W3C API.

**Pull request merged**